classes/Renderer/User.php

```diff
@@ -0,0 +1,178 @@
+            $editclass = '';
+        }
+
+        // The 'Log out' link.
+        $menudata   = $this->pagedata->page_metadata('userlogout', 'menu');
+        $logouttext = $menudata['text'];
+        $logouttitle= $menudata['title'];
+
+        $LOGOUTURL  = new \URL('userlogout');
+        if ($this->page != 'userlogout') {
+            $LOGOUTURL->insert(array("ret"=>$this->returl));
+            $logoutclass = '';
+        } else {
+            $logoutclass = 'on';
+        }
+
+        $username = $this->user->firstname() . ' ' . $this->user->lastname();
+
+        $this->data['user_nav_links'][] = array(
+            'href'    => $LOGOUTURL->generate(),
+            'title'   => $logouttitle,
+            'classes' => $logoutclass,
+            'text'    => $logouttext
+        );
+        $this->data['user_nav_links'][] = array(
+            'href'    => $EDITURL->generate(),
+            'title'   => $edittitle,
+            'classes' => $editclass,
+            'text'    => $edittext
+        );
+        $this->data['user_nav_links'][] = array(
+            'href'    => $EDITURL->generate(),
+            'title'   => $edittitle,
+            'classes' => $editclass,
+            'text'    => _htmlentities($username)
+        );
+    }
+
+    private function addLoggedOutLinks() {
+        // The 'Join' link.
+        $menudata   = $this->pagedata->page_metadata('userjoin', 'menu');
+        $jointext   = $menudata['text'];
+        $jointitle  = $menudata['title'];
+
+        $JOINURL    = new \URL('userjoin');
+        if ($this->page != 'userjoin') {
+            if ($this->page != 'userlogout' && $this->page != 'userlogin') {
+                // We don't do this on the logout page, because then the user
+                // will return straight to the logout page and be logged out
+                // immediately!
+                $JOINURL->insert(array("ret"=>$this->returl));
+            }
+            $joinclass = '';
+        } else {
+            $joinclass = 'on';
+        }
+
+        // The 'Log in' link.
+        $menudata   = $this->pagedata->page_metadata('userlogin', 'menu');
+        $logintext  = $menudata['text'];
+        $logintitle = $menudata['title'];
+
+        $LOGINURL   = new \URL('userlogin');
+        if ($this->page != 'userlogin') {
+            if ($this->page != "userlogout" &&
+                $this->page != "userpassword" &&
+                $this->page != 'userjoin') {
+                // We don't do this on the logout page, because then the user
+                // will return straight to the logout page and be logged out
+                // immediately!
+                // And it's also silly if we're sent back to Change Password.
+                // And the join page.
+                $LOGINURL->insert(array("ret"=>$this->returl));
+            }
+            $loginclass = '';
+        } else {
+            $loginclass = 'on';
+        }
+
+            $this->data['user_nav_links'][] = array(
```

Comment:
Spacing goes a bit wrong here (perhaps as this is a new file also get rid of the extra spaces either side of `=` and so on).